File: app/src/main/java/com/theappsolutions/boilerplate/ui/auth/AuthActivity.java
package com.theappsolutions.boilerplate.ui.auth;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.TextInputLayout;
import android.view.KeyEvent;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.TextView;

import com.theappsolutions.boilerplate.R;
import com.theappsolutions.boilerplate._additional_useful_classes.ui.menu.BottomMenuActivity;
import com.theappsolutions.boilerplate.custom_views.CustomToolbar;
import com.theappsolutions.boilerplate.other.change_log.ChangeLogManager;
import com.theappsolutions.boilerplate.ui.base.BasePresenter;
import com.theappsolutions.boilerplate.ui.projects.ProjectListActivity;
import com.theappsolutions.boilerplate.util.other_utils.permissions_check.PermissionsCheckActivity;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;
import butterknife.OnEditorAction;
import butterknife.OnTextChanged;

/**
 * @author Severyn Parkhomenko s.parkhomenko@theappsolutions.com
 * @copyright (c) 2018 TheAppSolutions. (https://theappsolutions.com)
 */
public class AuthActivity extends PermissionsCheckActivity implements AuthView {

    @BindView(R.id.et_login)
    EditText edtLogin;

    @BindView(R.id.et_password)
    EditText edtPassword;

    @BindView(R.id.login_input_lay)
    TextInputLayout textInputLayLogin;

    @BindView(R.id.pass_input_lay)
    TextInputLayout textInputLayPassword;

    @Inject
    AuthPresenter authPresenter;

    public static Intent getStartIntent(Context context) {
        return new Intent(context, AuthActivity.class);
    }

    public static void start(@NonNull Activity activity) {
        Intent intent = getStartIntent(activity);
        activity.startActivity(intent);
    }


    public static void startClearTop(@NonNull Activity activity) {
        Intent intent = new Intent(activity, AuthActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        activity.startActivity(intent);
    }

    @Override
    public void goAheadAfterPermissionsEnabling() {
        ChangeLogManager manager = new ChangeLogManager(this);
        manager.analyze();

        authPresenter.attachView(this);
    }

    @OnEditorAction(R.id.et_password)
    public boolean onEtPassAction(TextView v, int actionId, KeyEvent event) {
        if (actionId == EditorInfo.IME_ACTION_DONE) {
            onLogInButtonClick();
        }
        return true;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityComponent().inject(this);
        processPermissionAfterStart();
    }

    @OnClick(R.id.btn_login)
    public void onLogInButtonClick() {
        String login = edtLogin.getText().toString();
        String password = edtPassword.getText().toString();
        authPresenter.tryLogIn(login, password);
    }

    @Override
    protected int getContentViewRes() {
        return R.layout.activity_auth;
    }

    @Override
    protected CustomToolbar.Config getToolbarConfig() {
        return CustomToolbar.Config.withoutToolbar();
    }

    @Override
    public void showLoginError(String text) {
        textInputLayLogin.setError(text);
    }

    @Override
    public void showPasswordError(String text) {
        textInputLayPassword.setError(text);
    }

    @OnTextChanged(R.id.et_login)
    void onLoginTextChanged(CharSequence text) {
        textInputLayLogin.setError(null);
    }

    @OnTextChanged(R.id.et_password)
    void onPassTextChanged(CharSequence text) {
        textInputLayPassword.setError(null);
    }

    @Override
    public void navigateToMenuScreen() {
        BottomMenuActivity.start(this);
    }

    @Override
    public void navigateToListScreen() {
        ProjectListActivity.startWithAnimation(this);
    }

    @Override
    public BasePresenter getBasePresenter() {
        return authPresenter;
    }

}

File: app/src/main/java/com/theappsolutions/boilerplate/ui/auth/AuthActivity.kt
package com.theappsolutions.boilerplate.ui.auth

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.TextInputLayout
import android.view.KeyEvent
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.TextView
import butterknife.OnTextChanged

import com.theappsolutions.boilerplate.R
import com.theappsolutions.boilerplate.additionalusefulclasses.ui.menu.BottomMenuActivity
import com.theappsolutions.boilerplate.customviews.CustomToolbar
import com.theappsolutions.boilerplate.other.changelog.ChangeLogManager
import com.theappsolutions.boilerplate.ui.base.BasePresenter
import com.theappsolutions.boilerplate.ui.projects.ProjectListActivity
import com.theappsolutions.boilerplate.util.other.permissions_check.PermissionsCheckActivity

import javax.inject.Inject


import com.theappsolutions.boilerplate.TasBoilerplateApplication
import kotlinx.android.synthetic.main.activity_auth.*

/**
 * @author Severyn Parkhomenko s.parkhomenko@theappsolutions.com
 * @copyright (c) 2018 TheAppSolutions. (https://theappsolutions.com)
 */
class AuthActivity : PermissionsCheckActivity(), AuthView {

    @Inject
    lateinit var authPresenter: AuthPresenter

    override fun goAheadAfterPermissionsEnabling() {
        val manager = ChangeLogManager(this)
        manager.analyze()

        authPresenter.attachView(this)
        et_password.setOnEditorActionListener { v, actionId, event ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                onLogInButtonClick()
            }
            true
        }

        btn_login.setOnClickListener {
            onLogInButtonClick()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activityComponent()?.inject(this);
        processPermissionAfterStart()
    }

    fun onLogInButtonClick() {
        val login = et_login?.text.toString()
        val password = et_password?.text.toString()
        authPresenter.tryLogIn(login, password)
    }

    override fun getContentViewRes(): Int {
        return R.layout.activity_auth
    }

    override fun getToolbarConfig(): CustomToolbar.Config? {
        return CustomToolbar.Config.withoutToolbar()
    }

    override fun showLoginError(text: String) {
        login_input_lay.error = text
    }

    override fun showPasswordError(text: String) {
        pass_input_lay.error = text
    }

    @OnTextChanged(R.id.et_login)
    internal fun onLoginTextChanged(text: CharSequence) {
        login_input_lay.error = null
    }

    @OnTextChanged(R.id.et_password)
    internal fun onPassTextChanged(text: CharSequence) {
        pass_input_lay.error = null
    }

    override fun navigateToMenuScreen() {
        BottomMenuActivity.start(this)
    }

    override fun navigateToListScreen() {
        ProjectListActivity.startWithAnimation(this)
    }

    override fun getBasePresenter(): BasePresenter<*> {
        return authPresenter
    }

    companion object {

        fun getStartIntent(context: Context): Intent {
            return Intent(context, AuthActivity::class.java)
        }

        fun start(activity: Activity) {
            val intent = getStartIntent(activity)
            activity.startActivity(intent)
        }

        fun startClearTop(activity: Activity) {
            val intent = Intent(activity, AuthActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            activity.startActivity(intent)
        }
    }

}
